plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectorTest.java

```diff
@@ -1671,4 +1685,204 @@ protected TestTable simpleTable()
     {
         return new TestTable(onRemoteDatabase(), format("%s.simple_table", getSession().getSchema().orElseThrow()), "(col BIGINT)", ImmutableList.of("1", "2"));
     }
+
+    @DataProvider
+    public Object[][] fixedJoinDistributionTypes()
+    {
+        return new Object[][] {{BROADCAST}, {PARTITIONED}};
+    }
+
+    @Test(timeOut = 60_000, dataProvider = "fixedJoinDistributionTypes")
+    public void testDynamicFiltering(JoinDistributionType joinDistributionType)
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_DYNAMIC_FILTER_PUSHDOWN));
+        assertDynamicFiltering(
+                "SELECT * FROM orders a JOIN orders b ON a.orderkey = b.orderkey AND b.totalprice < 1000",
+                joinDistributionType,
+                hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN));
+    }
+
+    @Test(timeOut = 60_000)
+    public void testDynamicFilteringWithAggregationGroupingColumn()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_DYNAMIC_FILTER_PUSHDOWN));
+        assertDynamicFiltering(
+                "SELECT * FROM (SELECT orderkey, count(*) FROM orders GROUP BY orderkey) a JOIN orders b " +
+                        "ON a.orderkey = b.orderkey AND b.totalprice < 1000",
+                PARTITIONED,
+                hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN));
+    }
+
+    @Test(timeOut = 60_000)
+    public void testDynamicFilteringWithAggregationAggregateColumn()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_DYNAMIC_FILTER_PUSHDOWN));
+        MaterializedResultWithQueryId resultWithQueryId = getDistributedQueryRunner()
+                .executeWithQueryId(getSession(), "SELECT custkey, count(*) count FROM orders GROUP BY custkey");
+        // Detecting whether above aggregation is fully pushed down explicitly as there are cases where SUPPORTS_AGGREGATION_PUSHDOWN
+        // is false as not all aggregations are pushed down but the above aggregation is.
+        boolean isAggregationPushedDown = getPhysicalInputPositions(resultWithQueryId.getQueryId()) == 1000;
+        assertDynamicFiltering(
+                "SELECT * FROM (SELECT custkey, count(*) count FROM orders GROUP BY custkey) a JOIN orders b " +
+                        "ON a.count = b.custkey AND b.totalprice < 1000",
+                PARTITIONED,
+                isAggregationPushedDown && hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN));
```

Comment:
why `hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN)` is needed?

Reply:
Some connectors don't support predicate pushdown, this is needed to run test successfully on them.

Reply:
Can we make `hasBehavior(SUPPORTS_PREDICATE_PUSHDOWN)` part of skip conditions?

Why is the pushdown required? DF shouldn't differ regardless if DF is pushed or not.

Reply:
DF becomes part of regular filter before query to the external source, so it can't do any filtering of physical input positions if connector does not perform predicate pushdown.
I did not make it part of skip conditions because it's still useful to do a sanity test that DF isn't doing something unexpected.

Reply:
> DF becomes part of regular filter before query to the external source, so it can't do any filtering of physical input positions if connector does not perform predicate pushdown.

Connector shouldn't set `SUPPORTS_DYNAMIC_FILTER_PUSHDOWN` when more basic component required (`SUPPORTS_PREDICATE_PUSHDOWN`) does not work

Reply:
I've simplified this to set default value of SUPPORTS_DYNAMIC_FILTER_PUSHDOWN in BaseJdbcConnectorTest to SUPPORTS_PREDICATE_PUSHDOWN